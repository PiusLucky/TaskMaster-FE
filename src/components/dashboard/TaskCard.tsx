"use client";

import React, { useEffect, useState } from "react";
import { Badge } from "../ui/badge";
import { Avatar, AvatarFallback } from "@/components/ui/avatar";
import {
  Popover,
  PopoverContent,
  PopoverTrigger,
} from "@/components/ui/popover";
import { Task, UserData } from "@/types/global-types";
import { capitalizeFirstLetter, ellipisifyString, getFirstLetters } from "@/lib/utils";
import { TaskModel } from "../modals/TaskModal";
import { cookieStorageManager } from "@/api";
import { storageKeys } from "@/api/storageKeys";
import DeleteModal from "../modals/DeleteModal";
import { PriorityType } from "@/lib/enum";


function TaskCard({
  task,
  rerenderParentComp,
}: {
  task: Task;
  rerenderParentComp: () => void;
}) {
  const [isPopOverOpen, setIsPopOverOpen] = useState(false);

  const user = cookieStorageManager.getItem<UserData>(storageKeys.user);

  return (
    <section className="rounded-[0.75rem] shadow-lg pl-[1.5rem] pr-[1rem] pt-[1.25rem] pb-[1.69rem]">
      <div className="flex justify-between items-center">
        <div>
          <Badge
            className={`bg-[#24975633] hover:bg-[#24975633] text-[#219653] `}
          >
            <span className="">{capitalizeFirstLetter(task.category)}</span>
          </Badge>
        </div>
        <div>
          <Popover open={isPopOverOpen} onOpenChange={setIsPopOverOpen}>
            <PopoverTrigger>
              <img src="/svgs/three_dot.svg" alt="three dots for action" />
            </PopoverTrigger>
            <PopoverContent>
              <div className="text-[#2C3E50] flex flex-col gap-[0.8rem]">
                <TaskModel
                  rerenderParentComp={rerenderParentComp}
                  isCreate={false}
                  task={task}
                />

                <DeleteModal
                  taskId={task?.id}
                  title={task?.title}
                  setIsPopOverOpen={setIsPopOverOpen}
                  rerenderFunc={rerenderParentComp}
                />
              </div>
            </PopoverContent>
          </Popover>
        </div>
      </div>
      <div className="mt-[0.81rem] pr-2">{ellipisifyString(task.title, 100)}</div>
      <div className="mt-[0.81rem] pr-2 text-gray-500">{ellipisifyString(task.description, 300)}</div>
      <div className="flex gap-[2.5rem] mt-[0.81rem]">
        <div className="flex gap-2 items-center">
          <img src="/svgs/calendar.svg" alt="calendar icon" className="w-4" />
          <p className="text-[#828282] text-[0.75rem]">
            Due Date: {task.dueDate.toString()}
          </p>
        </div>

        {task.priority === PriorityType.High ? (
          <Badge
            className={`bg-[#e9081533] hover:bg-[#e9081533] text-[#FD0404] h-[1.437rem] !w-[150px] flex justify-center`}
          >
            <span className="text-[0.625rem]">
              {capitalizeFirstLetter(task.priority)} priority
            </span>
          </Badge>
        ) : (
          ""
        )}

        {task.priority === PriorityType.Low ? (
          <Badge
            className={`bg-[#27D17F] hover:bg-[#27D17F] text-[#fff] h-[1.437rem] !w-[150px] flex justify-center`}
          >
            <span className="text-[0.625rem]">
              {capitalizeFirstLetter(task.priority)} priority
            </span>
          </Badge>
        ) : (
          ""
        )}

        {task.priority === PriorityType.Medium ? (
          <Badge
            className={`bg-[#ffa50033] hover:bg-[#ffa50033] text-black h-[1.437rem] !w-[150px] flex justify-center`}
          >
            <span className="text-[0.625rem]">
              {capitalizeFirstLetter(task.priority)} priority
            </span>
          </Badge>
        ) : (
          ""
        )}

        <Avatar className="w-[1.5rem] h-[1.5rem]">
          <AvatarFallback className="text-[.8rem]">
            {getFirstLetters(user?.full_name || "TM")}
          </AvatarFallback>
        </Avatar>
      </div>
    </section>
  );
}

export default TaskCard;
