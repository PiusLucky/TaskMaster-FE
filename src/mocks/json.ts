import { CategoryType, PriorityType } from "@/lib/enum";

export const loginJson = {
  data: {
    access_token:
      "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJmcmVzaCI6ZmFsc2UsImlhdCI6MTY5NjA5MzQxMCwianRpIjoiMzI2ZTUwMDYtYzEzYy00OWUzLThmY2EtNmVmOWNmMDFlNzXXXXXXX",
    user_id: "3ecc04bf-1518-4b59-915f-4ba13a26df13",
  },
  meta: {
    message: "Success",
    statusCode: 200,
  },
};

export const registerJson = {
  data: {
    email: "lucddkypiussdd5cx0@ss.com",
    full_name: "Lucky Pius",
    id: "65731a4c-6322-4e99-a76d-b38afe71f223",
  },
  meta: {
    message: "Success",
    statusCode: 200,
  },
};

export const taskJson = {
  category: CategoryType.Fun,
  description:
    "Edin Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
  dueDate: new Date("Sun, 10 Dec 2023 00:00:00 GMT"),
  id: "3b84eb5a-1179-47da-99ef-f3d0ccae4c8b",
  priority: PriorityType.Medium,
  title: "Smith",
  user_id: "3ecc04bf-1518-4b59-915f-4ba13a26df13",
};

export const userJson = {
  data: {
    email: "xxxx@ss.com",
    full_name: "Lucky Pius",
    id: "3ecc04bf-1518-4b59-915f-4ba13a26df13",
  },
  meta: {
    message: "Success",
    statusCode: 200,
  },
};

export const taskListJson = {
  data: {
    pagination: {
      limit: 20,
      page: 1,
      total_elements: 1,
      total_pages: 1,
    },
    tasks: [{ ...taskJson }],
  },
  meta: {
    message: "Success",
    statusCode: 200,
  },
};

export const singleTaskJson = {
  data: {
    ...taskJson,
  },
  meta: {
    message: "Success",
    statusCode: 200,
  },
};
