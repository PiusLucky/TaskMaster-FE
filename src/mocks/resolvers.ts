import { loginJson } from "./json";

export const loginResolver = (req: any, res: any, ctx: any) => {
  return res(ctx.status(200), ctx.json(loginJson));
};
