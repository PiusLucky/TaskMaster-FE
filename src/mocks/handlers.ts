import { rest } from "msw";
import { loginResolver } from "./resolvers";
import apiResources from "@/api/resources";

// Matches any "GET /user" requests,
// and responds using the `responseResolver` function.
export const handlers = [
  rest.get(apiResources.user + "/users/login", loginResolver),
];
