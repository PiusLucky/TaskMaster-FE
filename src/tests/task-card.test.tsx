import { expect, test } from "vitest";
import { render } from "@testing-library/react";
import renderer from "react-test-renderer";
import TaskCard from "@/components/dashboard/TaskCard";
import { taskJson } from "@/mocks/json";

const rerenderParentComp = () => {};

test("Should be able to see the category badge", async () => {
  const { getByTestId } = render(
    <TaskCard task={taskJson} rerenderParentComp={rerenderParentComp} />
  );

  const topSection = getByTestId("category-badge");
  expect(topSection).to.exist;
});

test("Should be able to see the task title", async () => {
  const { getByTestId } = render(
    <TaskCard task={taskJson} rerenderParentComp={rerenderParentComp} />
  );

  const feature = getByTestId("title");
  expect(feature).to.exist;
});

test("Should be able to see the task description", async () => {
  const { getByTestId } = render(
    <TaskCard task={taskJson} rerenderParentComp={rerenderParentComp} />
  );

  const feature = getByTestId("description");
  expect(feature).to.exist;
});

test("Should be able to see the task due date", async () => {
  const { getByTestId } = render(
    <TaskCard task={taskJson} rerenderParentComp={rerenderParentComp} />
  );

  const feature = getByTestId("due-date");
  expect(feature).to.exist;
});

test("Should be able to see the priority badge", async () => {
  const { getByTestId } = render(
    <TaskCard task={taskJson} rerenderParentComp={rerenderParentComp} />
  );

  const feature = getByTestId("priority-badge");
  expect(feature).to.exist;
});

//Snapshot Testing
it("TaskCard should match snapshots", () => {
  const tree = renderer
    .create(
      <TaskCard task={taskJson} rerenderParentComp={rerenderParentComp} />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
