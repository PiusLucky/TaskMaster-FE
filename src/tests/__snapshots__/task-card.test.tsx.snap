// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TaskCard should match snapshots 1`] = `
<section
  className="rounded-[0.75rem] shadow-lg pl-[1.5rem] pr-[1rem] pt-[1.25rem] pb-[1.69rem]"
>
  <div
    className="flex justify-between items-center"
  >
    <div>
      <div
        className="inline-flex items-center rounded-full border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-[#24975633] hover:bg-[#24975633] text-[#219653]"
        data-testid="category-badge"
      >
        <span
          className=""
        >
          Fun
        </span>
      </div>
    </div>
    <div>
      <button
        aria-controls="radix-:r0:"
        aria-expanded={false}
        aria-haspopup="dialog"
        data-state="closed"
        onClick={[Function]}
        type="button"
      >
        <img
          alt="three dots for action"
          src="/svgs/three_dot.svg"
        />
      </button>
    </div>
  </div>
  <div
    className="mt-[0.81rem] pr-2"
    data-testid="title"
  >
    Smith
  </div>
  <div
    className="mt-[0.81rem] pr-2 text-gray-500"
    data-testid="description"
  >
    Edin Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum....
  </div>
  <div
    className="flex gap-[2.5rem] mt-[0.81rem]"
    data-testid="due-date"
  >
    <div
      className="flex gap-2 items-center"
    >
      <img
        alt="calendar icon"
        className="w-4"
        src="/svgs/calendar.svg"
      />
      <p
        className="text-[#828282] text-[0.75rem]"
      >
        Due Date: 
        Sun Dec 10 2023 01:00:00 GMT+0100 (West Africa Standard Time)
      </p>
    </div>
    <div
      data-testid="priority-badge"
    >
      <div
        className="items-center rounded-full border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-[#ffa50033] hover:bg-[#ffa50033] text-black h-[1.437rem] !w-[150px] flex justify-center"
      >
        <span
          className="text-[0.625rem]"
        >
          Medium
           priority
        </span>
      </div>
    </div>
    <span
      className="relative flex shrink-0 overflow-hidden rounded-full w-[1.5rem] h-[1.5rem]"
    >
      <span
        className="flex h-full w-full items-center justify-center rounded-full bg-muted text-[.8rem]"
      >
        T
      </span>
    </span>
  </div>
</section>
`;
